File: src/enmon/reading.processor.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import { Inject, Injectable, Logger } from '@nestjs/common';
import { Define, Queue } from 'agenda-nest';
import { EnmonApiClient } from './ApiClient.js';
import { UploadReading } from './upload-reading.schema.js';
import { UploadReadingRepository } from './upload-reading.repository.js';
import { READINGS_QUEUE_NAME, UPLOAD_JOB_NAME } from './constants.js';
import { UploadErrorFilter } from './upload-error.filter.js';

@Injectable()
@Queue(READINGS_QUEUE_NAME)
export class ReadingProcessor {
  private readonly logger = new Logger(ReadingProcessor.name);

  constructor(
    @Inject(EnmonApiClient)
    private enmonApiClient: EnmonApiClient,
    @Inject(UploadReadingRepository)
    private uploadDataRepository: UploadReadingRepository,
    @Inject(AsyncLocalStorage)
    private als: AsyncLocalStorage<{ readingId: unknown }>,
    @Inject(UploadErrorFilter)
    private uploadErrorFilter: UploadErrorFilter,
  ) {}

  @Define(UPLOAD_JOB_NAME)
  async handleUploadJob() {
    this.logger.log('processing job...');

    const cursor = this.uploadDataRepository.getSorterCursor();

    // eslint-disable-next-line no-restricted-syntax
    for await (const data of cursor) {
      await this.als.run({ readingId: data._id }, async () => {
        this.logger.log({ message: 'uploading reading...', id: data._id });

        try {
          await this.uploadReading(data);

          // eslint-disable-next-line no-underscore-dangle
          this.logger.log('deleting uploaded reading...');
          await data.deleteOne();
        } catch (e) {
          this.uploadErrorFilter.catch(e);
        }
      });
    }

    this.logger.log('all readings uploaded');
  }

  private async uploadReading({ config, reading }: UploadReading) {
    const { env, customerId, devEUI, token } = config;

    const payload = {
      devEUI,
      date: reading.readAt,
      value: reading.value,
      meterRegister: reading.register,
    } as const;

    this.logger.log({ message: 'uploading reading ...', payload });

    const { status, statusText } = await this.enmonApiClient.postMeterPlainValue({
      env,
      customerId,
      token,
      payload,
    });

    this.logger.log({ message: 'reading uploaded', status, statusText });
  }
}

File: dotnet/App/Enmon/ReadingProcessor.cs
using Hangfire;
using Microsoft.Extensions.Logging;
using MongoDB.Bson;
using MongoDB.Entities;

namespace Enmon;

[Queue(Constants.ReadingsQueueName)]
public class ReadingProcessor(
    ILogger<ReadingProcessor> logger,
    ApiClient enmonApiClient,
    IUploadJobQueue uploadJobQueue)
{
  [AutomaticRetry(Attempts = 3)]
  public async Task UploadReadingsAsync(CancellationToken cancellationToken)
  {
    logger.LogInformation("Processing upload job...");

    var cursor = await uploadJobQueue.GetSorterCursorAsync();

    while (await cursor.MoveNextAsync())
    {
      foreach (var document in cursor.Current)
      {
        cancellationToken.ThrowIfCancellationRequested();
        try
        {
          await UploadReadingAsync(document, cancellationToken);
          logger.LogInformation("Deleting uploaded reading...");
          await document.DeleteAsync();
        }
        catch (OperationCanceledException)
        {
          throw;
        }
        catch (Exception ex)
        {
          HandleUploadReadingError(ex);
        }
      }
    }

    logger.LogInformation("All readings uploaded.");
  }

  private async Task UploadReadingAsync(UploadReading data, CancellationToken cancellationToken)
  {
    var config = data.Config;
    var reading = data.Reading;

    var payload = new PlainDataPoint
    {
      DevEUI = config.DevEUI,
      Date = reading.ReadAt,
      Value = reading.Value,
      MeterRegister = reading.Register
    };

    logger.LogInformation("Uploading reading with payload: {Payload}", payload);

    try
    {
      var response = await enmonApiClient.PostMeterPlainValue(new PostMeterPlainValueArgs
      {
        Env = config.Env,
        CustomerId = new ObjectId(config.CustomerId),
        Token = config.Token,
        Payload = payload
      }, cancellationToken);
      logger.LogInformation("Reading uploaded successfully. Status: {StatusCode}, StatusText: {ReasonPhrase}", response.StatusCode, response.ReasonPhrase);
    }
    catch (HttpRequestException ex)
    {
      logger.LogError("HTTP Request failed: {Message}", ex.Message);
      throw;
    }
  }

  private void HandleUploadReadingError(Exception ex)
  {
    if (ex is HttpRequestException httpRequestException)
    {
      logger.LogError("Upload reading failed. Status: {StatusCode}, Message: {Message}", httpRequestException.StatusCode, httpRequestException.Message);
    }
    else
    {
      logger.LogError("An unexpected error occurred: {Message}", ex.Message);
      throw ex;
    }
  }
}
